src/ukify/ukify.py

```diff
@@ -0,0 +1,476 @@
+
+    def __post_init__(self):
+        self.offset = round_to_page(pe_executable_size(self.executable))
+
+    def add_section(self, section):
+        assert self.offset
+        assert section.offset is None
+        section.offset = self.offset
+        self.offset += round_to_page(section.size())
+        self.sections += [section]
+
+
+def parse_banks(s):
+    banks = s.split(',')
+    # TODO: do some sanity checking here
+    return banks
+
+
+def check_splash(filename):
+    if filename is None:
+        return
+
+    # import is delayed, to avoid import when the splash image is not used
+    try:
+        from PIL import Image
+    except ImportError:
+        return
+
+    img = Image.open(filename)
+    print(f'Splash image {filename} is {img.width}×{img.height} pixels')
+
+
+def check_inputs(opts):
+    for name, value in vars(opts).items():
+        if name in {'output', 'tools'}:
+            continue
+
+        if not isinstance(value, pathlib.Path):
+            continue
+
+        # Open file to check that we can read it, or generate an exception
+        value.open().close()
+
+    check_splash(opts.splash)
+
+
+def find_tool(name, fallback=None, opts=None):
+    if opts and opts.tools:
+        tool = opts.tools / name
+        if tool.exists():
+            return tool
+
+    return fallback or name
+
+
+def make_uki(opts):
+    # kernel payload signing
+
+    sbsign_tool = find_tool('sbsign', opts=opts)
+    sbsign_invocation = [
+        sbsign_tool,
+        '--key', opts.sb_key,
+        '--cert', opts.sb_cert,
+    ]
+
+    sign_kernel = opts.sign_kernel
+    if sign_kernel is None and opts.sb_key:
+        # figure out if we should sign the kernel
+        sbverify_tool = find_tool('sbverify', opts=opts)
+
+        cmd = [
+            sbverify_tool,
+            '--list',
+            opts.linux,
+        ]
+
+        print('+', shell_join(cmd))
+        info = subprocess.check_output(cmd, text=True)
+
+        # sbverify has wonderful API
+        if 'No signature table present' in info:
+            sign_kernel = True
+
+    if sign_kernel:
+        linux_signed = tempfile.NamedTemporaryFile(prefix='linux-signed')
+        linux = linux_signed.name
+
+        cmd = [
+            *sbsign_invocation,
+            opts.linux,
+            '--output', linux,
+        ]
+
+        print('+', shell_join(cmd))
+        subprocess.check_call(cmd)
+    else:
+        linux = opts.linux
+
+    uki = UKI(opts.stub)
+
+    # TODO: derive public key from from opts.pcr_private_key?
+    pcrpkey = opts.pcrpkey or opts.pcr_public_key
+
+    sections = [
+        # name,     content,         measure?
+        ('osrel',   opts.os_release, True ),
+        ('cmdline', opts.cmdline,    True ),
+        ('dtb',     opts.devicetree, True ),
+        ('splash',  opts.splash,     True ),
+        ('pcrpkey', pcrpkey,         True ),
+        ('initrd',  opts.initrd,     True ),
+        ('uname',   opts.uname,      False),
```

Comment:
this should be measured, really.

Reply:
(i mean, a patch adding measurement of this to sd-stub is trivial, might include in some PR)

Reply:
OK. But let's leave it for later…